### PR TITLE
USIO: Save Game Settings - Full Implementation (Supplement to #12686)

### PR DESCRIPTION
**Pull request by brian218**
This PR implemented permanent USIO game data saving, serving as a supplement to #12686.

### CODE REVIEW

File: rpcs3/Emu/Io/usio.cpp [hunk 1]
```diff
@@ -80,10 +81,12 @@ usb_device_usio::usb_device_usio(const std::array<u8, 7>& location)
 
 	g_fxo->get<usio_memory>().backup_memory.resize(0xB8);
 	g_fxo->get<usio_memory>().last_game_status.resize(0x28);
+	load_backup(g_cfg_vfs.get_device(g_cfg_vfs.dev_usb, "/dev_usb000").path + "usiobackup.bin");
```
Comment (elad335) [inline rpcs3/Emu/Io/usio.cpp:84]: 1. Must work with entwined launches of different games.
2. Must not touch PS3 filesystem.

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:84]: > 1. Must work with entwined launches of different games.
> 
>     2. Must not touch PS3 filesystem.

@elad335
1. Only a few games that adopt USIO save game data in this way. Any other ordinary PS3 games won't be affected by this at all.
2. Any suggestion about where to store the backup file?

Thank you.

Comment (elad335) [reply rpcs3/Emu/Io/usio.cpp:84]: How does the ps3 do it?

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:84]: > How does the ps3 do it?

I'm not pretty clear about what you are referring to.
In this PR, the backup file will be stored at and read from /dev_usb000/usiobackup.bin

Comment (elad335) [reply rpcs3/Emu/Io/usio.cpp:84]: I mean how the real ps3 save it?

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:84]: > I mean how the real ps3 save it?

A real USIO board has an on-board battery (just like the CMOS battery, which is used to keep BIOS settings, on your PC's motherboard), and it uses the power from the battery to keep data in its built-in SRAM.
The data of this PR's usiobackup.bin is originally stored in the USIO board's SRAM.

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:84]: > So it's shared with different games on real ps3? If so maybe save it inside /dev_hdd1/caches/somefile. This directory is not actually attached by itself to ps3 filesystem.

Yes. In addition, this kind of games (which adopt a USIO board and save their data in this way) all share the same Title ID "SCEEXE000"; so is it fine to save the backup file at _dev_hdd1/caches/SCEEXE000_SCEEXE000/usiobackup.bin_?

Comment (elad335) [reply rpcs3/Emu/Io/usio.cpp:84]: Yes, just dev_hdd1/caches/usiobackup.bin.

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Updated.

File: rpcs3/Emu/Io/usio.cpp [hunk 2]
```diff
@@ -102,6 +105,46 @@ void usb_device_usio::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue
 
 extern bool is_input_allowed();
 
+void usb_device_usio::load_backup(const std::string& path)
+{
+	fs::file backup_file(path, fs::create + fs::read + fs::write + fs::lock);
```
Comment (elad335) [inline rpcs3/Emu/Io/usio.cpp:110]: Use usio_backup_file.open.

File: rpcs3/Emu/Io/usio.cpp [hunk 3]
```diff
@@ -102,6 +105,46 @@ void usb_device_usio::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue
...
+		return;
+	}
+
+	std::vector<u8> buffer(0xE0);
+	memcpy(buffer.data(), g_fxo->get<usio_memory>().backup_memory.data(), 0xB8);
+	memcpy(buffer.data() + 0xB8, g_fxo->get<usio_memory>().last_game_status.data(), 0x28);
+
+	usio_backup_file.seek(0, fs::seek_set);
+	usio_backup_file.trunc(buffer.size());
+	usio_backup_file.write(buffer.data(), buffer.size());
```
Comment (elad335) [inline rpcs3/Emu/Io/usio.cpp:144]: Sikpler to make it 2 writes.

File: rpcs3/Emu/Io/usio.cpp [hunk 4]
```diff
@@ -102,6 +105,46 @@ void usb_device_usio::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue
...
+	usio_backup_file.close();
```
Comment (elad335) [inline rpcs3/Emu/Io/usio.cpp:145]: This line is not needed because it's part of fs::file dtor.

**Review by elad335 — CHANGES_REQUESTED** [review]

File: rpcs3/Emu/Io/usio.cpp (hunk 4 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Done.

File: rpcs3/Emu/Io/usio.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/Emu/Io/usio.cpp:142]: fs::file::write does not guarentee atomicity, for that you need fs::pending_file.

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:142]: > fs::file::write does not guarantee atomicity, for that you need fs::pending_file.

Why can't we just use trunc(buffer.size()) to resize the file first and then write(buffer.data(), buffer.size())?

File: rpcs3/Emu/Io/usio.cpp [hunk 5]
```diff
@@ -102,6 +105,46 @@ void usb_device_usio::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue
...
+
+	std::vector<u8> buffer(0xE0);
+	if (usio_backup_file.size() != buffer.size())
+	{
+		usio_log.notice("Invalid USIO Backup file detected. Treating it as an empty file: %s", path);
+		return;
+	}
+
+	usio_backup_file.read(buffer.data(), buffer.size());
+	memcpy(g_fxo->get<usio_memory>().backup_memory.data(), buffer.data(), 0xB8);
```
[the remark this answers is not in the document]
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp:125]: > 2 reads is fine.

What do you mean by this?

[the remark this answers is not in the document]
Comment (elad335) [reply rpcs3/Emu/Io/usio.cpp:125]: I don't see why not reading directly onto the two vectors.

[the remark this answers is not in the document]
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Thanks a lot for suggestion!

**Comment by Nekotekina** [comment]
I suppose it should be written to file immediately once any changes are made

**Comment by brian218** [comment]
> I suppose it should be written to file immediately once any changes are made

Wouldn't this cause performance compromise during the game?
The game writes data to these two buffers (backup_memory and last_game_status) frequently no matter whether a change is made by a player.

**Comment by RipleyTom** [comment]
> > I suppose it should be written to file immediately once any changes are made
> 
> Wouldn't this cause performance compromise during the game? The game writes data to these two buffers (backup_memory and last_game_status) frequently no matter whether a change is made by a player.

You could just compare size and content and if it changes write to file?

**Comment by brian218** [comment]
> > > I suppose it should be written to file immediately once any changes are made
> > 
> > 
> > Wouldn't this cause performance compromise during the game? The game writes data to these two buffers (backup_memory and last_game_status) frequently no matter whether a change is made by a player.
> 
> You could just compare size and content and if it changes write to file?

@RipleyTom The sizes of backup_memory(0xB8) and last_game_status(0x28) are constant, so the total size is always be 0xE0(0xB8 + 0x28).
We cannot discriminate changes by comparing their sizes.

As for the content, since a timestamp is included, the game changes the content and writes it to the endpoints frequently.

**Comment by Nekotekina** [comment]
Hmm, perhaps a memory mapped file can be used as an alternative.

**Comment by brian218** [comment]
> Hmm, perhaps a memory mapped file can be used as an alternative.

@Nekotekina Could you suggest an example? Thank you.
By the way, if frequent writing to the backup file won't compromise the performance during the game, I can of course make it write changes to the backup file each time any change is made by the game.

**Comment by elad335** [comment]
File writes can be async if this function changes content so frequantly.

**Comment by Nekotekina** [comment]
Actually nevermind, it would need quite a bit of work. I can implement it later. PR can be merged as is.

**Comment by brian218** [comment]
> Actually nevermind, it would need quite a bit of work. I can implement it later. PR can be merged as is.

@Nekotekina Thanks!

**Pull request merged**